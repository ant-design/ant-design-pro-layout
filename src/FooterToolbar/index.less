@import '~antd/es/style/themes/default.less';

@pro-footer-bar-prefix-cls: ~'@{ant-prefix}-pro-footer-bar';

.@{pro-footer-bar-prefix-cls} {
  position: fixed;
  right: 0;
  bottom: 0;
  z-index: 99;
  display: flex;
  width: 100%;
  padding: 0 24px;
  line-height: 48px;
  background: @component-background;
  border-top: 1px solid @border-color-split;
  box-shadow: @box-shadow-base;
  &-left {
    flex: 1;
  }
}
