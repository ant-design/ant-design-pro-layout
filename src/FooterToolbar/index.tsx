import React, { useContext, useMemo, ReactNode } from 'react';
import { RouteContext, RouteContextType } from '../index';
import classNames from 'classnames';
import './index.less';
import { Space } from 'antd';

export interface FooterToolbarProps {
  extra?: React.ReactNode;
  style?: React.CSSProperties;
  className?: string;
  renderContent?: (
    props: FooterToolbarProps & RouteContextType & { leftWidth?: string },
    dom: JSX.Element,
  ) => ReactNode;
  prefixCls?: string;
}
const FooterToolbar: React.FC<FooterToolbarProps> = (props) => {
  const {
    children,
    prefixCls = 'ant-pro',
    className,
    extra,
    renderContent,
    ...restProps
  } = props;

  const baseClassName = `${prefixCls}-footer-bar`;
  const value = useContext(RouteContext);
  const width = useMemo(() => {
    const { hasSiderMenu, isMobile, siderWidth } = value;
    if (!hasSiderMenu) {
      return undefined;
    }
    // 0 or undefined
    if (!siderWidth) {
      return '100%';
    }
    return isMobile ? '100%' : `calc(100% - ${siderWidth}px)`;
  }, [value.collapsed, value.hasSiderMenu, value.isMobile, value.siderWidth]);

  const dom = (
    <>
      <div className={`${baseClassName}-left`}>{extra}</div>
      <div className={`${baseClassName}-right`}>
        <Space>{children}</Space>
      </div>
    </>
  );

  return (
    <div
      className={classNames(className, `${baseClassName}`)}
      style={{ width }}
      {...restProps}
    >
      {renderContent
        ? renderContent(
            {
              ...props,
              ...value,
              leftWidth: width,
            },
            dom,
          )
        : dom}
    </div>
  );
};

export default FooterToolbar;
