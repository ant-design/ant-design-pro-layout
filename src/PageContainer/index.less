@import '~antd/es/style/themes/default.less';

@pro-layout-page-container: ~'@{ant-prefix}-pro-page-container';

.@{pro-layout-page-container}-children-content {
  margin: 24px 24px 0;
}

.@{pro-layout-page-container} {
  &-warp {
    background-color: @component-background;
    .@{ant-prefix}-tabs-nav {
      margin: 0;
    }
  }
  &-ghost {
    .@{pro-layout-page-container}-warp {
      background-color: transparent;
    }
  }
}

.@{pro-layout-page-container}-main {
  .@{pro-layout-page-container}-detail {
    display: flex;
  }

  .@{pro-layout-page-container}-row {
    display: flex;
    width: 100%;
  }

  .@{pro-layout-page-container}-title-content {
    margin-bottom: 16px;
  }

  .@{pro-layout-page-container}-title,
  .@{pro-layout-page-container}-content {
    flex: auto;
  }

  .@{pro-layout-page-container}-extraContent,
  .@{pro-layout-page-container}-main {
    flex: 0 1 auto;
  }

  .@{pro-layout-page-container}-main {
    width: 100%;
  }

  .@{pro-layout-page-container}-title {
    margin-bottom: 16px;
  }

  .@{pro-layout-page-container}-logo {
    margin-bottom: 16px;
  }

  .@{pro-layout-page-container}-extraContent {
    min-width: 242px;
    margin-left: 88px;
    text-align: right;
  }
}

@media screen and (max-width: @screen-xl) {
  .@{pro-layout-page-container}-main {
    .@{pro-layout-page-container}-extraContent {
      margin-left: 44px;
    }
  }
}

@media screen and (max-width: @screen-lg) {
  .@{pro-layout-page-container}-main {
    .@{pro-layout-page-container}-extraContent {
      margin-left: 20px;
    }
  }
}

@media screen and (max-width: @screen-md) {
  .@{pro-layout-page-container}-main {
    .@{pro-layout-page-container}-row {
      display: block;
    }

    .@{pro-layout-page-container}-action,
    .@{pro-layout-page-container}-extraContent {
      margin-left: 0;
      text-align: left;
    }
  }
}

@media screen and (max-width: @screen-sm) {
  .@{pro-layout-page-container}-detail {
    display: block;
  }
  .@{pro-layout-page-container}-extraContent {
    margin-left: 0;
  }
}
