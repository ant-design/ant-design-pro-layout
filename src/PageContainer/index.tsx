import { PageHeader, Tabs } from 'antd';
import React, { useContext, ReactNode } from 'react';
import classNames from 'classnames';
import { TabsProps, TabPaneProps } from 'antd/es/tabs';
import { PageHeaderProps } from 'antd/es/page-header';
import './index.less';
import RouteContext, { RouteContextType } from '../RouteContext';
import GridContent from '../GridContent';
import FooterToolbar from '../FooterToolbar';

export interface PageHeaderTabConfig {
  tabList?: TabPaneProps[];
  tabActiveKey?: TabsProps['activeKey'];
  onTabChange?: TabsProps['onChange'];
  tabBarExtraContent?: TabsProps['tabBarExtraContent'];
  tabProps?: TabsProps;
}

export interface PageContainerProps
  extends PageHeaderTabConfig,
    Omit<PageHeaderProps, 'title'> {
  title?: React.ReactNode | false;
  content?: React.ReactNode;
  extraContent?: React.ReactNode;
  prefixCls?: string;
  footer?: ReactNode[];
  pageHeaderRender?: (props: PageContainerProps) => React.ReactNode;
}

/**
 * render Footer tabList
 * In order to be compatible with the old version of the PageHeader
 * basically all the functions are implemented.
 */
const renderFooter: React.SFC<Omit<
  PageContainerProps & {
    prefixedClassName: string;
  },
  'title'
>> = ({
  tabList,
  tabActiveKey,
  onTabChange,
  tabBarExtraContent,
  tabProps,
  prefixedClassName,
}) => {
  if (tabList && tabList.length) {
    return (
      <Tabs
        className={`${prefixedClassName}-tabs`}
        activeKey={tabActiveKey}
        onChange={(key) => {
          if (onTabChange) {
            onTabChange(key);
          }
        }}
        tabBarExtraContent={tabBarExtraContent}
        {...tabProps}
      >
        {tabList.map((item, index) => (
          // eslint-disable-next-line react/no-array-index-key
          <Tabs.TabPane {...item} tab={item.tab} key={item.key || index} />
        ))}
      </Tabs>
    );
  }
  return null;
};

const renderPageHeader = (
  content: React.ReactNode,
  extraContent: React.ReactNode,
  prefixedClassName: string,
): React.ReactNode => {
  if (!content && !extraContent) {
    return null;
  }
  return (
    <div className={`${prefixedClassName}-detail`}>
      <div className={`${prefixedClassName}-main`}>
        <div className={`${prefixedClassName}-row`}>
          {content && (
            <div className={`${prefixedClassName}-content`}>{content}</div>
          )}
          {extraContent && (
            <div className={`${prefixedClassName}-extraContent`}>
              {extraContent}
            </div>
          )}
        </div>
      </div>
    </div>
  );
};

const defaultPageHeaderRender = (
  props: PageContainerProps,
  value: RouteContextType & { prefixedClassName: string },
): React.ReactNode => {
  const {
    title,
    content,
    pageHeaderRender,
    extraContent,
    style,

    prefixCls,
    ...restProps
  } = props;

  if (pageHeaderRender) {
    return pageHeaderRender({ ...props, ...value });
  }
  let pageHeaderTitle = title;
  if (!title && title !== false) {
    pageHeaderTitle = value.title;
  }
  return (
    <PageHeader
      {...value}
      title={pageHeaderTitle}
      {...restProps}
      footer={renderFooter({
        ...restProps,
        prefixedClassName: value.prefixedClassName,
      })}
      prefixCls={prefixCls}
    >
      {renderPageHeader(content, extraContent, value.prefixedClassName)}
    </PageHeader>
  );
};

const PageContainer: React.FC<PageContainerProps> = (props) => {
  const { children, style, footer, prefixCls = 'ant-pro' } = props;
  const value = useContext(RouteContext);

  const prefixedClassName = `${prefixCls}-page-container`;

  const className = classNames(prefixedClassName, props.className);

  return (
    <div style={style} className={className}>
      <div className={`${prefixedClassName}-warp`}>
        {defaultPageHeaderRender(props, {
          ...value,
          prefixCls: undefined,
          prefixedClassName,
        })}
      </div>
      <GridContent>
        {children ? (
          <div>
            <div className={`${prefixedClassName}-children-content`}>
              {children}
            </div>
            {footer && (
              <div
                style={{
                  height: 48,
                  marginTop: 24,
                }}
              />
            )}
          </div>
        ) : null}
      </GridContent>
      {footer && <FooterToolbar>{footer}</FooterToolbar>}
    </div>
  );
};

export default PageContainer;
