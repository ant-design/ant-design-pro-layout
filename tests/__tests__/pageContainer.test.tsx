import { render } from 'enzyme';

import React from 'react';
import PageContainer from '../../src/PageContainer';
import FooterToolbar from '../../src/FooterToolbar';

describe('PageContainer', () => {
  beforeAll(() => {
    Object.defineProperty(window, 'matchMedia', {
      value: jest.fn(() => ({
        matches: false,
        addListener() {},
        removeListener() {},
      })),
    });
  });

  it('💄 base use', async () => {
    const html = render(<PageContainer title="期贤" />);
    expect(html).toMatchSnapshot();
  });

  it('⚡️ support footer', async () => {
    const html = render(
      <PageContainer
        title="期贤"
        footer={[<button type="button">right</button>]}
      />,
    );
    expect(html).toMatchSnapshot();
  });

  it('🔥 support footer', async () => {
    const html = render(
      <PageContainer
        title="期贤"
        breadcrumb={{
          routes: [
            {
              path: '/',
              breadcrumbName: 'home',
            },
          ],
        }}
        footer={[
          <button key="right" type="button">
            right
          </button>,
        ]}
      />,
    );
    expect(html).toMatchSnapshot();
  });

  it('🔥footer bar support extra', async () => {
    const html = render(
      <FooterToolbar
        className="qixian_footer"
        extra={
          <img
            src="https://gw.alipayobjects.com/zos/rmsportal/KDpgvguMpGfqaHPjicRK.svg"
            alt="logo"
          />
        }
      >
        <button type="button">right</button>
      </FooterToolbar>,
    );
    expect(html).toMatchSnapshot();
  });

  it('🔥footer bar support renderContent', async () => {
    const html = render(
      <FooterToolbar
        className="qixian_footer"
        extra={
          <img
            src="https://gw.alipayobjects.com/zos/rmsportal/KDpgvguMpGfqaHPjicRK.svg"
            alt="logo"
          />
        }
        renderContent={() => {
          return 'home_toolbar';
        }}
      >
        <button type="button">right</button>
      </FooterToolbar>,
    );
    expect(html).toMatchSnapshot();
  });
});
