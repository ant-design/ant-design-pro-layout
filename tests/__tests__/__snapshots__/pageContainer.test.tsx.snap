// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PageContainer ⚡️ support footer 1`] = `
<div
  class="ant-pro-page-container"
>
  <div
    class="ant-pro-page-container-warp"
  >
    <div
      class="ant-page-header ant-page-header-ghost"
    >
      <div
        class="ant-page-header-heading"
      >
        <div
          class="ant-page-header-heading-left"
        >
          <span
            class="ant-page-header-heading-title"
            title="期贤"
          >
            期贤
          </span>
        </div>
      </div>
    </div>
  </div>
  <div
    class="ant-pro-grid-content"
  >
    <div
      class="ant-pro-grid-content-children"
    />
  </div>
  <div
    class="ant-pro-footer-bar"
  >
    <div
      class="ant-pro-footer-bar-left"
    />
    <div
      class="ant-pro-footer-bar-right"
    >
      <div
        class="ant-space ant-space-horizontal ant-space-align-center"
      >
        <div
          class="ant-space-item"
        >
          <button
            type="button"
          >
            right
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`PageContainer 💄 base use 1`] = `
<div
  class="ant-pro-page-container"
>
  <div
    class="ant-pro-page-container-warp"
  >
    <div
      class="ant-page-header ant-page-header-ghost"
    >
      <div
        class="ant-page-header-heading"
      >
        <div
          class="ant-page-header-heading-left"
        >
          <span
            class="ant-page-header-heading-title"
            title="期贤"
          >
            期贤
          </span>
        </div>
      </div>
    </div>
  </div>
  <div
    class="ant-pro-grid-content"
  >
    <div
      class="ant-pro-grid-content-children"
    />
  </div>
</div>
`;

exports[`PageContainer 🔥 support footer 1`] = `
<div
  class="ant-pro-page-container"
>
  <div
    class="ant-pro-page-container-warp"
  >
    <div
      class="ant-page-header has-breadcrumb ant-page-header-ghost"
    >
      <div
        class="ant-breadcrumb"
      >
        <span>
          <span
            class="ant-breadcrumb-link"
          >
            <span>
              home
            </span>
          </span>
          <span
            class="ant-breadcrumb-separator"
          >
            /
          </span>
        </span>
      </div>
      <div
        class="ant-page-header-heading"
      >
        <div
          class="ant-page-header-heading-left"
        >
          <span
            class="ant-page-header-heading-title"
            title="期贤"
          >
            期贤
          </span>
        </div>
      </div>
    </div>
  </div>
  <div
    class="ant-pro-grid-content"
  >
    <div
      class="ant-pro-grid-content-children"
    />
  </div>
  <div
    class="ant-pro-footer-bar"
  >
    <div
      class="ant-pro-footer-bar-left"
    />
    <div
      class="ant-pro-footer-bar-right"
    >
      <div
        class="ant-space ant-space-horizontal ant-space-align-center"
      >
        <div
          class="ant-space-item"
        >
          <button
            type="button"
          >
            right
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`PageContainer 🔥footer bar support extra 1`] = `
<div
  class="qixian_footer ant-pro-footer-bar"
>
  <div
    class="ant-pro-footer-bar-left"
  >
    <img
      alt="logo"
      src="https://gw.alipayobjects.com/zos/rmsportal/KDpgvguMpGfqaHPjicRK.svg"
    />
  </div>
  <div
    class="ant-pro-footer-bar-right"
  >
    <div
      class="ant-space ant-space-horizontal ant-space-align-center"
    >
      <div
        class="ant-space-item"
      >
        <button
          type="button"
        >
          right
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`PageContainer 🔥footer bar support renderContent 1`] = `
<div
  class="qixian_footer ant-pro-footer-bar"
>
  home_toolbar
</div>
`;
